### ISSUE TITLE
Implement background process manager

### ISSUE THREAD

**Issue by stgraber** [lxc/lxd#5826]
LXD currently spawns a number of background processes and then attempts to track them as they run. This isn't done very consistently and is the source of quite a bit of code duplication.

Instead we should add a new package that provides functions to both spawn a background process and keep track of it as well as import existing background processes and keep track of those too.

The process struct should then let us stop, start, restart, reload, signal, wait for the process.
When spawning a process, we will need to provide a way to specify the environment and process ownership information so we can properly respawn the process should it die.

The process monitoring logic should auto-respawn processes up to a limit within a reasonable time period and when available should make use of pidfds as a race free way of keeping track of those processes.

Initial users should be:
 - forkdns
 - dnsmasq
 - forkproxy

**Comment by ribsthakkar** [comment]
Hi @stgraber, my team with @abbykrish, @anusha-paul, and myself will work on this issue

**Comment by stgraber** [comment]
@ribsthakkar assigned it to you for now, I'll need to have the other two comment so Github lets me assign it to them too.

**Comment by stgraber** [comment]
For this issue, we'd like a completely standalone Go package that we could then transition some of our code over. So rather than directly modifying LXD from the start, this should start as a new Go package under our `shared/` directory, defining the struct and functions to achieve the goals outlined in this issue.

Once we think it covers all we need to spawn/stop/restart/track background processes on Linux, we can add some unit tests for it, clean the package up to be of the same standard as our newer standalone packages and then have separate commits or even a separate branch making use of it to drive `dnsmasq` and `forkdns` as two of the easier users of this.

Down the line, we'll want to make this use `pidfs` but that's I believe out of scope for the current chunk of work and something that @brauner and myself should think about later on.

**Comment by stgraber** [comment]
For this one, good contacts would be @brauner, @tomponline and myself

**Comment by ribsthakkar** [comment]
Thank you for the updated information. We'll let you know about updates when we get to working on them!

**Comment by abbykrish** [comment]
Hi @stgraber, I'm Abby, could you assign me to the issue? Thanks ahead!

**Comment by anusha-paul** [comment]
Hi @stgraber, could you add me to this issue as well?

**Comment by stgraber** [comment]
done!

**Comment by ribsthakkar** [comment]
Hi all @stgraber @tomponline @brauner 
We are starting to spec out this this issue. We have approximately two and a half weeks before we need to present some sort of running implementation of this issue. With that in mind, we're hoping to get a more clear understanding of what should be the key milestone(s) for this issue. I don't expect this package to be fully fleshed given the time frame and the learning curve so far with the first issue 😅, but I hope we can get something basic up and running before the deadline.

If I am understanding this correctly, we would essentially be creating a package to wrap around system calls such as fork and exec to create new processes? However, the package itself would store/manage the processes and PIDs? What should be the first main tasks?

**Comment by ribsthakkar** [comment]
Also, is this a process manager for processes that the user wants to trigger inside containers or that LXD creates internally?

**Comment by ribsthakkar** [comment]
Another question, does the package keep track of all of the processes that have spawned or the package just returns a process struct that can be used by lxd to spawn, kill etc ?

**Comment by ribsthakkar** [comment]
Just following up again @stgraber @tomponline @brauner

**Comment by stgraber** [comment]
Sorry about the delay.

We're looking for a package which would allow for:
 - NewProcess(name string, args... string) (*Process, error)
 - ImportProcess(path string) (*Process, error)

With Process supporting:
 - Pid() (int64, error)
 - Stop() error
 - Start() error
 - Restart() error
 - Reload() error
 - Save(path string) error
 - Signal(int64) error
 - Wait() (int64, error)

We'll likely want a number of attributes on Process to configure things like the specific signals to send for stop, restart, reload, ..., configuration of environment, stdin/stdout/stderr setup, ...

But the general idea is that we can spawn a subprocess using this, then keep track of it and interact with it through the struct directly or if we lost track of the struct, reload it from disk. This is effectively a fancier version of a PID file which allows re-importing all the state needed to restart or reload the process.

**Comment by ribsthakkar** [comment]
I see. Essentially we'd be storing the same thing a pidfile (which is just the pid anyway) does but along with that the configuration options of the process, path to executable executable, the arguments, etc?
I suppose the first iteration of this would be a simplified implementation of Python's subprocess module?

**Comment by stgraber** [comment]
Yeah, it's effectively a fancy pid file, I'd recommend storing as yaml since that's what we tend to do in LXD. The functions above should be pretty easy to map to `exec.Command` and `exec.Command.Process`.

**Comment by ribsthakkar** [comment]
What is the functional difference between restarting a process and reloading a process? I suppose restarting would be to kill the current process if it is running and start it again, but I am not sure about reloading?

**Comment by stgraber** [comment]
Reload is usually sending a signal like SIGHUP to cause the process to reload its configuration.

**Comment by ribsthakkar** [comment]
Hi Stephane, so we have the functionality you requested implemented? We are working on writing Go Tests for it. Where should we be putting the test file(s)?

**Comment by stgraber** [comment]
If using go tests, you should just use the default _test.go files so they're picked up and run by 'go test'.

The whole package will most likely reside under 'lxd/subprocess' with the package similarly called subprocess (bit of a call back to the similar Python module).

**Comment by ribsthakkar** [comment]
Should we be writing dedicated unit tests for each of the functions or a few tests that cover all of the functions?

**Comment by stgraber** [comment]
I would go with a few tests that cover all the functions, will be easier to reason about and update in the future.

**Comment by abbykrish** [comment]
Hi Stephane! We finished writing the subprocess package with the functionalities you had mentioned. It is in PR #6544

One thing we wanted to point out -- for the sake of easy testing, we wrote Python scripts to be the processes that we managed in the tests. Is this okay? We had some difficulty with Go scripts because we weren't able to declare them in package `main` because they were within our subprocess folder. If there's an easy way to get around this issue, we'd be more than happy to write Go scripts seeing as there is no .py files in LXD at the moment.

**Comment by stgraber** [comment]
I'll have to take a closer look. Spawning subprocesses from go test doesn't seem trivial and is likely too heavy weight. But maybe we can at least switching from python to a shell script to reduce the dependency chain for those tests.

**Comment by stgraber** [comment]
Keeping this issue open as we still need to port `dnsmasq` and `forkdns` over to the new background process manager code before we can call the issue done.

**Comment by ribsthakkar** [comment]
Sounds good. I will be out of town for the next couple weeks, but I will get to working on this after I come back.

**Comment by stgraber** [comment]
@ribsthakkar excellent, thanks!

**Comment by techtonik** [comment]
I thought that reusing something like https://github.com/ochinchina/supervisord should be easier, but ok.

**Comment by stgraber** [comment]
Not really, we'd only want a very very very small subset of what this implements and we wanted this to be under our control so we can easily integrate it with specific Linux kernel features we've been implementing very recently (`pidfds`) without having to care about external users, backward compatibility or other operating systems.

**Comment by ribsthakkar** [comment]
Just browsing the code, I'd be modifying the dnsmasq.go file and main_forkdns.go correct?

**Comment by ribsthakkar** [comment]
Aside from the kill functionality in dnsmasq.go, I can't really seem to find where else this would integrate

**Comment by stgraber** [comment]
I believe we both kill dnsmasq during a restart or when modifying some files and we also occasionally tell it to reload, unless we got rid of that somehow.

So I'd have expected us to modify `network.go` to spawn dnsmasq through the new management code, serialize that down to disk as a replacement of the pid file, then load that back up and use it for reload/kill where we currently rely on `dnsmasq.pid`.

**Comment by ribsthakkar** [comment]
I see, there is also the spawnForkDNS in the network.go file, I presume it would have the same kind of treatment?

**Comment by stgraber** [comment]
Yep

**Comment by ribsthakkar** [comment]
I did a bit of the refactoring, let me know if I was on the right track

**Comment by stgraber** [comment]
@ribsthakkar thanks but the branch doesn't appear to be building right now :(
